### DISCUSSION TITLE
Behaviour of ABORT / QUIT with respect to CATCH / THROW

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#120] [Q&A]
Forth 200x Draft 19.1 makes no mention of the behaviour of `ABORT` and `QUIT` (exception ext) with respect to `CATCH` and `THROW`.  If you consider the supplied reference implementations for E.9.6.1.0875 CATCH and E.9.6.1.2275 THROW, either `ABORT` or `QUIT` will empty the return stack (only changing the RSP typically) discarding any catch frames, but fail to clear the catch `HANDLER` variable.

A subsequent `THROW` would attempt to restore the stacks from `HANDLER` but either stack's contents will likely have been altered and impossible to restore correctly.

Similarly when a new `CATCH` is set, the saved frame will save `HANDLER` and start working from the bottom of the return stack altering contents.  Several  attempt to `THROW` will hit the bad `HANDLER` and result in undefined behaviour.

I believe `ABORT` and `QUIT` (exception ext) need to address this by stating that any pointer(s) to a catch frame (`HANDLER`) must also be cleared.

**Comment by MitchBradley** [comment]
And thus we see why QUIT really should be defined in terms of CATCH/THROW, instead of explicitly emptying the return stack.  History.

**Comment by ruv** [comment]
> I believe `ABORT` and `QUIT` (exception ext) need to address this by stating that any pointer(s) to a catch frame (`HANDLER`) must also be cleared.

The normative text doesn't use the notion of a _catch frame pointer_ at all, so it should not state that such pointers must be cleared.

The reference implementations are not normative. If a bug is found in a reference implementation (for example, if `HANDLER` is not cleared when it's required), it should be just fixed.


> A subsequent `THROW` would attempt to restore the stacks from `HANDLER` but either stack's contents will likely have been altered and impossible to restore correctly.

A comment to the `CATCH` and `THROW` reference implementation says: "This sample implementation does not explicitly handle the case in which `CATCH` has never been called (i.e., the `ABORT` behavior)".   So the mentioned case is just not considered in the reference implementation for the sake of simplicity.
